Modules, Exports, and Imports.md

```diff
@@ -0,0 +1,301 @@
+
+Along the way, we'll see both ways to structure code well, and ways to shoot yourself in the foot. We'll also learn more
+about Elm's tooling and `elm-package.json`.
+
+This shouldn't be the first guide on Elm you read, but if you've playing around for a few hours and have specific
+questions about modules and the like, you should be fine. This guide aims to be a comprehensive reference to be read
+straight through, but if you have questions about a specific piece of syntax you should do fine skimming for code
+blocks.
+
+## Packages and Applications
+
+Elm programs fall into two categories. **Packages** (or **libraries**) are meant to be reused and are typically
+published to the [package registry](http://package.elm-lang.org/). **Applications** are programs that can actually be
+run and produce output. Applications may depend on packages, which may depend on other packages in turn; applications
+are the root of this tree.
+
+The distinction between package and application occurs in `elm-package.json`: A package will export at least one module
+in the `"exposed-modules"` field. Packages are compiled with `elm make` and no additional arguments. This compiles the
+modules listed, and subjects them to additional compile-time checks ensuring every exported value and type are
+documented properly. The compiler does not generate an HTML or JS output file.
+
+If you pass a file to `elm make`, then you are writing an application. That file (and therefore module) is known as
+"Main". It's not necessary to call this file `Main.elm` or even declare a module definition in it at all -- but you
+should for large applications.
+
+However defined, only Main can define ports (a topic otherwise beyond the scope of this guide). The Main module must
+define a `main` value, whose type is restricted to a handful of things that the compiler can display. Other modules
+should avoid defining `main`.
+
+Another difference between packages and applications is how they tolerate change. Elm's package manager enforces
+[semantic versioning](http://semver.org/), discouraging breaking changes. Additionally anything that is exported must be
+documented. Packages therefore want to hide their type definitions and helper functions, in order to present a stable,
+understandable, and helpful interface. Applications are more tolerant of change; for example adding a field to a record
+type alias or a tag to a union type. Only the application itself relies on these definitions, and the compiler will find
+any discrepancies that changes introduce. Therefore, packages must tightly control what is exported, but applications
+can afford to be more permissive.
+
+## Modules and Files
+
+Every file contains exactly one module. Filenames should be capitalized and match the name of the module they contain.
+For example, `Foo.elm` should declare the module `Foo`. The compiler will not allow module names and file paths to
+disagree.
```

Comment:
Another example would be useful: `Foo/Bar.elm` should be declared as module `Foo.Bar`